src/Storages/HDFS/StorageHDFSCluster.h

```diff
@@ -39,11 +40,20 @@ class StorageHDFSCluster : public IStorage
 
     NamesAndTypesList getVirtuals() const override;
 
+    ClusterPtr getCluster(ContextPtr context) const override;
+    RemoteQueryExecutor::Extension getTaskIteratorExtension(ContextPtr context) const override;
+
 private:
     String cluster_name;
     String uri;
     String format_name;
     String compression_method;
+
+    mutable ClusterPtr cluster;
+    mutable std::shared_ptr<HDFSSource::DisclosedGlobIterator> iterator;
+    mutable std::shared_ptr<HDFSSource::IteratorWrapper> callback;
```

Comment:
Why mutable?
Create a struct with cluster, iterator, callback. Return it from createIteratorAndCallback.

Reply:
Because iterator and callback are different for different types of Storage (e.g. StorageHDFS and StorageS3). Maybe we can just remove `const` from method in base class